geopandas/geodataframe.py

```diff
@@ -1437,18 +1447,34 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwargs):
             func, axis=axis, raw=raw, result_type=result_type, args=args, **kwargs
         )
         # Reconstruct gdf if it was lost by apply
-        if (
-            isinstance(result, DataFrame)
-            and self._geometry_column_name in result.columns
-        ):
-            # axis=1 apply will split GeometryDType to object, try and cast back
+        if isinstance(result, DataFrame):
+            if self._geometry_column_name in result.columns:
+                # axis=1 apply will split GeometryDType to object, try and cast back
+                try:
+                    result = result.set_geometry(self._geometry_column_name)
+                except TypeError:
+                    pass
+                else:
+                    if self.crs is not None and result.crs is None:
+                        result.set_crs(self.crs, inplace=True)
+            # preserve other geom col dtypes
+            other_geom_cols = set(self.columns[self.dtypes == "geometry"]) - {
+                self._geometry_column_name
+            }
+            for c in other_geom_cols:
+                result.loc[:, c] = _ensure_geometry(result[c], crs=self[c].crs)
+            # if active geom col lost, but still have geometry
+            if type(result) == pd.DataFrame and (result.dtypes == "geometry").any():
+                result = GeoDataFrame(result, geometry=None)
```

Comment:
This section is to fix `test_apply_axis1_secondary_geo_cols` on pandas=1.0.5 (I was originally not going to implement this (https://github.com/geopandas/geopandas/pull/2332#issuecomment-1119544811)).

It has the side effect of also resolving this case which was a todo under `test_apply` (which is why it's not under a not PANDAS_GE11 guard. I might split this out to another PR to be clearer
```python
# TODO below should be a GeoDataFrame to be consistent with new getitem logic
#   leave as follow up as quite complicated
#   FrameColumnApply.series_generator returns object dtypes Series, so will have
#   patch result of apply
assert_object(df[["geometry2", "value1"]].apply(identity, axis=1), pd.DataFrame)
```

Reply:
> I might split this out to another PR to be clearer

+1

And will take a look at the rest of the PR now!

Reply:
@m-richards would it be possible to remove it here? (unless it depends on each other?)

I would like to focus on getting this one merged (with the idea that fixing the apply edge case is less of a priority)

Reply:
They are dependent, but only for a single case on pandas 1.0.5. I'll remove and add the xfail in two commits so you can clearly see what I mean.

Reply:
Thanks! I think it is fine to decide that a corner case only in 1.0.5 is not a high priority